File: bubbler_app/lib/data/repositories/preferences_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/preferences.dart';

/// Get / update recommendation preferences.
class PreferencesRepository {
  PreferencesRepository(this._client);

  final ApiClient _client;

  /// Current preferences (`GET /user/me/preferences`).
  Future<UserPreferences> getPreferences() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePreferences,
    );
    return UserPreferences.fromJson(_asJsonMap(data));
  }

  /// Replaces preferences (`PUT /user/me/preferences`).
  ///
  /// Callers should normalize [payload.strategyWeights] the same way as iOS
  /// before sending.
  Future<UserPreferences> updatePreferences(
    PreferencesUpdatePayload payload,
  ) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePreferences,
      method: 'PUT',
      data: payload.toJson(),
      contentType: 'application/json',
    );
    return UserPreferences.fromJson(_asJsonMap(data));
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/PreferencesRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.data.model.PreferencesUpdatePayload
import com.bubbler.android.data.model.UserPreferences
import okhttp3.RequestBody.Companion.toRequestBody

/** Get / update recommendation preferences. */
open class PreferencesRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun getPreferences(): UserPreferences =
        apiClient.get(Endpoints.USER_ME_PREFERENCES, token = tokenStore.requireAccessToken())

    open suspend fun updatePreferences(payload: PreferencesUpdatePayload): UserPreferences {
        val body = apiClient.json.encodeToString(
            PreferencesUpdatePayload.serializer(),
            payload,
        )
        return apiClient.request(
            path = Endpoints.USER_ME_PREFERENCES,
            method = "PUT",
            token = tokenStore.requireAccessToken(),
            body = body.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }
}
